interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,320 @@
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "mount.%s" element not a string (%q)`, fieldName, iface)
+			}
+			list = append(list, valueString)
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return errors.New(mountAttrNotADictionary)
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return errors.New(mountAttrNotADictionary)
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "mount.persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		// TODO: parse other fields
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
+		return fmt.Errorf(`mount-control "what" attribute must be an absolute path`)
+	}
+
+	if !cleanSubPath(what) {
+		return fmt.Errorf(`mount-control "what" pattern is not clean: %q`, what)
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
+	return nil
+}
+
+func validateWhereAttr(where string) error {
+	if !whereRegexp.MatchString(where) {
+		return fmt.Errorf(`mount-control "where" attribute is invalid`)
+	}
+
+	if !cleanSubPath(where) {
+		return fmt.Errorf(`mount-control "where" pattern is not clean: %q`, where)
+	}
+
+	if _, err := utils.NewPathPattern(where); err != nil {
+		return fmt.Errorf(`mount-control "where" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
+	return nil
+}
+
+func validateMountOptions(options []string) error {
+	for _, o := range options {
+		if !strutil.ListContains(allowedMountOptions, o) {
+			return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		}
+	}
+	return nil
+}
+
+func validateMountInfo(mountInfo *MountInfo) error {
+	if err := validateWhatAttr(mountInfo.what); err != nil {
+		return err
+	}
+
+	if err := validateWhereAttr(mountInfo.where); err != nil {
+		return err
+	}
+
+	if err := validateMountOptions(mountInfo.options); err != nil {
+		return err
+	}
+
+	// TODO: validate mount type
```

Comment:
Please add type validation (this could be similar to option validation where items are checked against an allow-list) so that we can ensure the generated apparmor profile cannot have malicious content inserted into it, ie if a type was specified as: `auto) options=() path/to/malicious/content /var/lib/snapd/hostfs/...,\n   mount fstype=(` - then I think currently this could allow a snap to mount over parts of the root file-system or some other trusted location, or they could inject other apparmor policy bits as well to gain additional accesses.

Reply:
I added a simple regular expression for alphanumeric characters, instead of a full list of possible options. This should address the security issue you raised, while remaining open to support any FS type. Please let me know, if you see a reason for an explicit list.